contracts/stellar-multicall/src/tests/test.rs

```diff
@@ -51,198 +59,413 @@ impl TestTarget {
-        FunctionCall {
-            contract: target.clone(),
-            function: symbol_short!("method"),
-            args: vec![&env, IntoVal::<_, Val>::into_val(&0u32, &env)],
-        },
+        construct_function_call!(
+            env,
+            gas_service_id,
+            owner,
+            pay_gas(
+                sender.to_val(),
+                destination_chain.to_val(),
+                destination_address.to_val(),
+                payload.to_val(),
+                spender.to_val(),
+                IntoVal::<_, Val>::into_val(&token, &env),
+                Bytes::new(&env).to_val()
+            )
+        ),
     ];
 
-    let test_client = TestTargetClient::new(&env, &target);
+    let transfer_token_auth = mock_auth!(
+        spender,
+        token_client.transfer(spender, gas_service_client.address, token.amount)
+    );
 
-    let transfer_ownership_auth =
-        mock_auth!(current_owner, test_client.transfer_ownership(&new_owner));
+    let pay_gas_auth = mock_auth!(
+        spender,
+        gas_service_client.pay_gas(
+            sender,
+            destination_chain,
+            destination_address,
+            payload,
+            spender,
+            token,
+            Bytes::new(&env)
+        ),
+        &[(transfer_token_auth.invoke).clone()]
+    );
+
+    let multicall_auth = mock_auth!(
+        owner,
+        client.multicall(&function_calls),
+        &[(pay_gas_auth.invoke).clone()]
+    );
 
     client
-        .mock_auths(&[transfer_ownership_auth])
+        .mock_auths(&[multicall_auth, pay_gas_auth])
         .multicall(&function_calls);
+
+    goldie::assert!(fmt_last_emitted_event::<GasPaidEvent>(&env));
+}
+
+#[test]
+fn multicall_zero_call_succeeds() {
+    let TestConfig { env, client, .. } = setup();
+
+    let function_calls = Vec::new(&env);
+
+    client.multicall(&function_calls);
 }
 
 #[test]
 #[should_panic(expected = "HostError: Error(Auth, InvalidAction)")]
-fn multicall_auth_no_auth_fails() {
-    let (env, client, target, _) = setup();
+fn multicall_no_auth_fails() {
+    let TestConfig {
+        env,
+        client,
+        target_id,
+        owner,
+    } = setup();
     let new_owner = Address::generate(&env);
 
     let function_calls = vec![
         &env,
-        FunctionCall {
-            contract: target.clone(),
-            function: Symbol::new(&env, "transfer_ownership"),
-            args: vec![&env, new_owner.to_val()],
-        },
-        FunctionCall {
-            contract: target,
-            function: symbol_short!("method"),
-            args: vec![&env, IntoVal::<_, Val>::into_val(&42u32, &env)],
-        },
+        construct_function_call!(
+            env,
+            target_id,
+            owner,
+            method(IntoVal::<_, Val>::into_val(&42u32, &env))
+        ),
+        construct_function_call!(
+            env,
+            target_id,
+            owner,
+            transfer_ownership(new_owner.to_val())
+        ),
```

Comment:
```suggestion
        function_call!(owner, client.transfer_ownership(new_owner)),
```
why can't this be made simpler just like `assert_auth!`?

https://github.com/axelarnetwork/axelar-cgp-stellar/blob/main/packages/stellar-axelar-std/src/error.rs#L171

Reply:
I applied the fixed as you described, however I had to leave in the `target_id` (contract ID) as a param because the `FunctionCall` needs it, and I don't think the contract ID is accessible from client

Reply:
`client.address`. look at the implementation of `assert_auth!` again, it doesn't have this problem